### PR TITLE
Drupal 8.6 support and entity class overriding support

### PR DESCRIPTION
**Pull request by tamasd**

### CODE REVIEW

File: tests/src/Functional/ApigeeEdgeTestTrait.php [hunk 1]
```diff
@@ -83,6 +84,7 @@ protected function invalidateKey() {
       'active_key_oauth_token' => '',
     ];
     $this->container->get('state')->set('apigee_edge.auth', $keys);
+    drupal_flush_all_caches();
```
Comment (mxr576) [inline tests/src/Functional/ApigeeEdgeTestTrait.php]: Clear states cache only with `$this->container->get('state')->resetCache()`

File: src/Entity/Controller/ApiProductController.php [hunk 2]
```diff
@@ -31,40 +29,11 @@
...
-    }
-    $this->entityClass = $entityClass;
-  }
-
   /**
    * {@inheritdoc}
    */
-  protected function getEntityClass(): string {
-    return $this->entityClass;
+  protected function getInterface(): string {
```
Comment (mxr576) [inline src/Entity/Controller/ApiProductController.php]: Let's call it `getEntityInterface()` instead.

File: src/Entity/Controller/DeveloperController.php [hunk 3]
```diff
@@ -21,57 +21,35 @@
...
-   *   The API client.
-   * @param string $entityClass
-   *   The FQCN of the entity class that is used in Drupal.
-   * @param \Apigee\Edge\Serializer\EntitySerializerInterface|null $entitySerializer
-   *   The entity serializer.
-   * @param \Apigee\Edge\Api\Management\Controller\OrganizationControllerInterface|null $organizationController
-   *   The organization controller.
-   *
-   * @throws \ReflectionException
+   * {@inheritdoc}
```
Comment (mxr576) [inline src/Entity/Controller/DeveloperController.php]: I would not use inheritdoc on constructor's description. Does PHPCS complains about the parameters if you do not provide a description to them?

File: src/Entity/Controller/DrupalEntityControllerAwareTrait.php [hunk 4]
```diff
@@ -31,16 +34,39 @@
...
 
+  /**
+   * @var string
+   */
+  protected $entityClass;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(string $organization, ClientInterface $client, string $entityClass, ?EntitySerializerInterface $entitySerializer = NULL, ?OrganizationControllerInterface $organizationController = NULL) {
```
Comment (mxr576) [inline src/Entity/Controller/DrupalEntityControllerAwareTrait.php]: Inheriting constructor from a trait? 😨`parent` does not point to anywhere in this context.

File: src/Entity/Controller/DrupalEntityControllerAwareTrait.php [hunk 5]
```diff
@@ -31,16 +34,39 @@
...
+    parent::__construct($organization, $client, $entitySerializer, $organizationController);
+    $rc = new \ReflectionClass($entityClass);
```
Comment (mxr576) [inline src/Entity/Controller/DrupalEntityControllerAwareTrait.php]: I'd rather create an utility function from this and call that in the constructor of the entity classes.

File: src/Entity/Controller/DrupalEntityControllerAwareTrait.php [hunk 6]
```diff
@@ -31,16 +34,39 @@
  */
 trait DrupalEntityControllerAwareTrait {
 
+  /**
+   * @var string
+   */
+  protected $entityClass;
```
Comment (mxr576) [inline src/Entity/Controller/DrupalEntityControllerAwareTrait.php]: Describe that this is an FQCN of an entity class.

File: src/Entity/Query/QueryFactory.php [hunk 7]
```diff
@@ -69,33 +69,107 @@ public function __construct(EntityTypeManagerInterface $entityTypeManager) {
...
+   * Otherwise the default \Drupal\apigee_edge\Entity\Query\Query class
+   * if being used.
+   *
+   * @param string $class
+   *   Entity class.
+   *
+   * @return null|string
+   *   Query class or null if not found.
+   */
+  protected function findQueryClass(string $class): ?string {
```
Comment (mxr576) [inline src/Entity/Query/QueryFactory.php]: If we would like to refactor this, can't we just introduce a new annotation in our entity classes that can defined the FQCN of the entity query class (with fallback to the default entity class)?

File: tests/src/Functional/ApigeeEdgeTestTrait.php [hunk 8]
```diff
@@ -72,6 +72,7 @@ protected function restoreKey() {
...
```
Comment (mxr576) [inline tests/src/Functional/ApigeeEdgeTestTrait.php]: Clear states cache only with `$this->container->get('state')->resetCache()`

File: tests/src/Functional/DeveloperAppFieldTest.php [hunk 9]
```diff
@@ -545,6 +546,7 @@ protected function assertFieldVisibleOnEntityForm(string $field_label, bool $vis
    */
   protected function assertFieldVisibleOnEntityDisplay(string $app_name, string $field_label, string $field_value, bool $visible = TRUE) {
     $this->drupalGet("/user/{$this->account->id()}/apps/{$app_name}");
+    $this->assertEquals(Response::HTTP_OK, $this->getSession()->getStatusCode());
```
Comment (mxr576) [inline tests/src/Functional/DeveloperAppFieldTest.php]: Why is this necessary?

**Review by mxr576 — CHANGES_REQUESTED** [review]

File: tests/src/Functional/ApigeeEdgeTestTrait.php (hunk 8 above)
Comment (tamasd) [reply tests/src/Functional/ApigeeEdgeTestTrait.php]: For some reason that doesn't do the trick in Drupal 8.6. I spent half a day trying to figure out why.

File: tests/src/Functional/DeveloperAppFieldTest.php (hunk 9 above)
Comment (tamasd) [reply tests/src/Functional/DeveloperAppFieldTest.php]: When the test fails here for some reason, I get a more meaningful message (e.g. 404 instead of 200), and not an "element xpath=blabla not found on page".

Comment (mxr576) [reply tests/src/Functional/DeveloperAppFieldTest.php]: Ok.

File: tests/src/Functional/ApigeeEdgeTestTrait.php (hunk 8 above)
Comment (mxr576) [reply tests/src/Functional/ApigeeEdgeTestTrait.php]: It would be really great of we could figure this out, because it is critical for us.

File: src/Entity/Controller/ApiProductController.php (hunk 2 above)
Comment (mxr576) [reply src/Entity/Controller/ApiProductController.php]: @tamasd  ^

**Comment by mxr576** [comment]
TODOs:
* Version information should be modified in README.md as well
* 8.5.x-dev references in [.travis.yml](.travis.yml) should be replaced with 8.6.x-dev.

@tamasd ^

File: src/Entity/Query/QueryFactory.php (hunk 7 above)
Comment (mxr576) [reply src/Entity/Query/QueryFactory.php]: @tamasd ^

File: src/Entity/Controller/DrupalEntityControllerAwareTrait.php [hunk 10]
```diff
@@ -31,16 +34,47 @@
...
+   * The FQCN of the Drupal entity class.
+   *
+   * @var string
+   */
+  protected $entityClass;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(string $organization, ClientInterface $client, string $entity_class, ?EntitySerializerInterface $entity_serializer = NULL, ?OrganizationControllerInterface $organization_controller = NULL) {
```
Comment (mxr576) [inline src/Entity/Controller/DrupalEntityControllerAwareTrait.php]: I think we should not inherit the constructor of a class from a trait.

**Review by mxr576 — CHANGES_REQUESTED** [review]

**Comment by mxr576** [comment]
Latest test results:
https://travis-ci.org/tamasd/apigee-edge-drupal/builds/426595129

**Review by mxr576 — APPROVED** [review]

**Review by peterserfozo — APPROVED** [review]
